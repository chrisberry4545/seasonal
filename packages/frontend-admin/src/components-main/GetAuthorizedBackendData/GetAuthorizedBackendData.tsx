import React, { FC, useState, useEffect } from 'react';
import { IDataFormConfigProps } from '../DataForm/DataForm';
import { LoadingSpinner } from '@chrisb-dev/seasonal-shared-frontend-components';
import './GetAuthorizedBackendData.scss';

export interface IGetAuthorizedBackendDataProps<T> {
  items: T;
  reload?: () => void;
  updateMethod?: (item: T) => Promise<T>;
  buttonText?: string;
  additionalConfig?: IDataFormConfigProps<T>;
}

interface IGetAuthorizedBackendDataState<T> {
  isLoading: boolean;
  items: T | null;
  error: string | null;
}

export function GetAuthorizedBackendData<T>({
  InnerComponent,
  requestDataMethod,
  updateMethod,
  buttonText,
  additionalConfig
}: {
  InnerComponent: FC<IGetAuthorizedBackendDataProps<T>>,
  requestDataMethod: () => Promise<T>,
  updateMethod?: (item: T) => Promise<T>,
  buttonText?: string,
  additionalConfig?: IDataFormConfigProps<T>
}) {

  const [state, setState] = useState<IGetAuthorizedBackendDataState<T>>({
    error: null,
    isLoading: true,
    items: null
  });

  const requestNewData = () => {
    requestDataMethod()
      .then((items) => setState({
        error: null,
        isLoading: false,
        items
      })).catch((error) => setState({
        error: error.message,
        isLoading: false,
        items: null
      }));
  };

  useEffect(requestNewData, []);

  return (
    <div>
      {
        state.isLoading
          ? <div className='c-get-authorized-backend-data__loading'>
            <LoadingSpinner />
          </div>
          : <div>
            {
              !state.error && state.items
                ? <InnerComponent
                    items={state.items}
                    reload={requestNewData}
                    updateMethod={updateMethod}
                    buttonText={buttonText}
                    additionalConfig={additionalConfig}
                  />
                : <div>{state.error}</div>
            }
          </div>
      }
    </div>
  );
}
