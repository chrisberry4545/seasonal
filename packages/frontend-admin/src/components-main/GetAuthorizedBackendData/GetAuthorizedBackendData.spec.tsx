import React from 'react';
import * as react from 'react';
import { ShallowWrapper, shallow } from 'enzyme';
import { GetAuthorizedBackendData } from './GetAuthorizedBackendData';
import { LoadingSpinner } from '@chrisb-dev/seasonal-shared-frontend-components';

jest.mock('@chrisb-dev/seasonal-shared-frontend-components', () => ({
  LoadingSpinner: () => 'LoadingSpinner'
}));

describe('GetAuthorizedBackendData', () => {
  let wrapper: ShallowWrapper;
  let mockInnerComponent: jest.Mock;
  let mockRequestDataMethod: jest.Mock;
  let mockUpdateMethod: jest.Mock;
  const additionalConfig = {};
  const items = [{}, {}];

  beforeEach(() => {
    mockInnerComponent = jest.fn();
    mockRequestDataMethod = jest.fn().mockResolvedValue(items);
    mockUpdateMethod = jest.fn();
  });

  describe('before the data has been retrieved', () => {
    beforeEach(() => {
      jest.spyOn(react, 'useEffect').mockImplementationOnce((f) => undefined);
      wrapper = shallow(
        <GetAuthorizedBackendData
          InnerComponent={mockInnerComponent}
          requestDataMethod={mockRequestDataMethod}
          updateMethod={mockUpdateMethod}
          additionalConfig={additionalConfig}
          />
      );
    });

    test('shows a loading spinner', () =>
      expect(wrapper.find(LoadingSpinner).exists()).toBe(true));

  });

  describe('after the data has been retrieved', () => {
    beforeEach(() => {
      jest.spyOn(react, 'useEffect').mockImplementationOnce((f) => f());
      wrapper = shallow(
        <GetAuthorizedBackendData
          InnerComponent={mockInnerComponent}
          requestDataMethod={mockRequestDataMethod}
          updateMethod={mockUpdateMethod}
          additionalConfig={additionalConfig}
          />
      );
    });

    test('renders correctly', () => expect(wrapper).toMatchSnapshot());

  });

  describe('if retrieving the data fails', () => {
    beforeEach(() => {
      jest.spyOn(react, 'useEffect').mockImplementationOnce((f) => f());
      mockRequestDataMethod = jest.fn().mockRejectedValue(new Error('error'));
      wrapper = shallow(
        <GetAuthorizedBackendData
          InnerComponent={mockInnerComponent}
          requestDataMethod={mockRequestDataMethod}
          updateMethod={mockUpdateMethod}
          additionalConfig={additionalConfig}
          />
      );
    });

    test('renders an error', () => expect(wrapper).toMatchSnapshot());
  });

});
