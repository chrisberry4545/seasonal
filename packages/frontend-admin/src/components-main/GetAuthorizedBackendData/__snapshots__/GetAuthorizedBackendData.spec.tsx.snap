// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GetAuthorizedBackendData after the data has been retrieved renders correctly 1`] = `
<div>
  <div>
    <mockConstructor
      additionalConfig={Object {}}
      items={
        Array [
          Object {},
          Object {},
        ]
      }
      reload={[Function]}
      updateMethod={[MockFunction]}
    />
  </div>
</div>
`;

exports[`GetAuthorizedBackendData if retrieving the data fails renders an error 1`] = `
<div>
  <div>
    <div>
      error
    </div>
  </div>
</div>
`;
